[Clang][CodeGen] Emit “trap reasons” on UBSan traps

anthonyhatran opened the pull request:
This patch adds a human readable trap category and message to UBSan traps. The category and message are encoded in a fake frame in the debug info where the function is a fake inline function where the name encodes the trap category and message. This is the same mechanism used by Clang’s `__builtin_verbose_trap()`.

This change allows consumers of binaries built with trapping UBSan to more easily identify the reason for trapping. In particular LLDB already has a frame recognizer that recognizes the fake function names emitted in debug info by this patch. A patch testing this behavior in LLDB will be added in a separate patch.

The human readable trap messages are based on the messages currently emitted by the userspace runtime for UBSan in compiler-rt. Note the wording is not identical because the userspace UBSan runtime has access to dynamic information that is not available during Clang’s codegen.

Test cases for each UBSan trap kind are included.

Complements the [`-fsanitize-annotate-debug-info` feature](https://github.com/llvm/llvm-project/pull/141997). While `-fsanitize-annotate-debug-info` attempts to annotate all UBSan-added instructions, this feature (`-fsanitize-debug-trap-reasons`) only annotates the final trap instruction using SanitizerHandler information.

This work is part of a GSoc 2025 project.

Credit to @delcypher for helping with the description.

anthonyhatran commented:
Anyways, to address Dan's comments, I changed all test cases so that they utilize `clang_cc1` instead of `clang`. This also meant that I had to change any `-fsanitize=undefined` and `-fsanitize-trap=undefined` to the specific check since `undefined` is a check grouping and is not recognized by front-end. I also adjusted the strings in the switch function to more closely match what was emitted when the test cases were run without `-fsanitize-trap=`, removed the unnecessary bug fix for unreachable, and added all test cases in for each `SanitizerHandler`, aside from `SanitizerHandler::BoundsSafety`.

Michael137 commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -85,6 +85,94 @@ enum VariableTypeDescriptorKind : uint16_t {
+    return "Signed integer divide or remainder overflowed";
+
+  case SanitizerHandler::DynamicTypeCacheMiss:
+    return "Dynamic-type cache miss";
```
I think we can improve the wording on this one. AFAIU this is something going wrong in the dynamic_cast runtime support

Michael137 reviewed (commented):
LGTM (modulo the question about one of the message strings)

I'll let @delcypher comment on whether all his concerns have been addressed

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -85,6 +85,94 @@ enum VariableTypeDescriptorKind : uint16_t {
+  case SanitizerHandler::VLABoundNotPositive:
+    return "Variable length array bound evaluates to non-positive value";
+
+  case SanitizerHandler::BoundsSafety:
```
Yes we should do that

delcypher commented on clang/test/CodeGen/ubsan-trap-reason-add-overflow.c:
```diff
@@ -0,0 +1,10 @@
+
+// CHECK: call void @llvm.ubsantrap(i8 0) {{.*}}!dbg [[LOC:![0-9]+]]
+// CHECK: [[LOC]] = !DILocation(line: 0, scope: [[MSG:![0-9]+]], {{.+}})
+// CHECK: distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer
```
The specific trap messages seem to be missing from the test cases.

anthonyhatran replied:
I explicitly left it out because I wanted to hear feedback on the strings after running tests without `-fsanitize-trap=undefined` and viewing what runtime UBSan (non-trapping) printed. I will add the TrapMessage checks in the next PR.

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4051,6 +4139,14 @@ void CodeGenFunction::EmitTrapCheck(llvm::Value *Checked,
+  llvm::DILocation *TrapLocation = Builder.getCurrentDebugLocation();
+  llvm::StringRef TrapMessage = GetUBSanTrapForHandler(CheckHandlerID);
+
+  if (getDebugInfo()) {
```
This looks wrong. If TrapMessage is empty we’ll attach a message (e.g. for the bounds safety trap) anyway.

delcypher reviewed (changes requested):
Looks like we need to fix a few more things. Did you resolve the issue with the existing test cases failing?

anthonyhatran commented:
> Looks like we need to fix a few more things. Did you resolve the issue with the existing test cases failing?

Apologies for the belated reply on this one; I ran the test suite overnight just to double-check that the test cases I listed were actually caused by my changes.

To answer your question,: yes, I have proposed solutions for most of them. Unfortunately I don't think there is just one adjustment that fixes all of them though (some include modifying the test cases themselves others involve changing what I added). I'll try to get the commit in as soon as possible so you can take a look.

anthonyhatran commented:
Commit #3 (338505e) Update:
Apologies for taking a bit on the commit, took a while to fix test cases. I'll list how the 5 test cases were fixed below:

1. {`clang/test/CodeGen/ubsan-trap-debugloc.c`}: Hand written, so I adjusted the unmerged trap checks to expect a synthetic `line: 0` trap-message location.
2. {`clang/test/CodeGen/bounds-checking-debuginfo.c, clang/test/CodeGen/cfi-icall-generalize-debuginfo.c, clang/test/CodeGen/cfi-icall-normalize2-debuginfo.c`}: Noticed that these were auto generated to match emitted IR. Since our feature made changes to emitted IR, I just reran utils/update_cc_test_checks.py to regenerate them.
3. {`clang/test/CodeGen/cfi-check-fail-debuginfo.c`}: This tried to wrap an already-synthetic location a second time causing a crash, so I added a guard in `EmitTrapCheck`: if the current `DILocation` is synthetic, we skip rewrapping it.

As for other changes, I re-added the `TrapMessage` is empty check, reworded `SanitizerHandler::DynamicTypeCacheMiss`'s string and updated the test cases to reflect the actual strings. Also added a comment for `BoundsSafety`

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -85,6 +85,96 @@ enum VariableTypeDescriptorKind : uint16_t {
+    return "Variable length array bound evaluates to non-positive value";
+
+  case SanitizerHandler::BoundsSafety:
+    // Not added as of current, will be added in future patch.
```
```suggestion
    // Explicitly not supported because `-fbounds-safety` is not part of UBSan
```

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4051,6 +4141,27 @@ void CodeGenFunction::EmitTrapCheck(llvm::Value *Checked,
+  llvm::StringRef TrapMessage = GetUBSanTrapForHandler(CheckHandlerID);
+
+  // Explicitly used to prevent infinite wrapping in cfi-check-fail-debuginfo.c
+  // test
```
I don't get this. How could we get infinite wrapping?

anthonyhatran replied:
Meant to say "double"

delcypher reviewed (changes requested):
I think this is pretty much at the point where we can start having other reviewers take a look at this. Let's address the minor comments, reword the PR description to describe what's implemented and then we can add other reviewers.

anthonyhatran commented:
Crash report from clang/test/CodeGen/cfi-check-fail-debuginfo.c:
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.    Program arguments: clang -cc1 -internal-isystem /usr/local/lib/clang/21/include -nostdsysteminc -triple x86_64-unknown-linux -O2 -fsanitize-cfi-cross-dso -fsanitize=cfi-icall,cfi-nvcall,cfi-vcall,cfi-unrelated-cast,cfi-derived-cast -fsanitize-trap=cfi-icall,cfi-nvcall -fsanitize-recover=cfi-vcall,cfi-unrelated-cast -fsanitize-annotate-debug-info=cfi-icall,cfi-nvcall,cfi-vcall,cfi-unrelated-cast,cfi-derived-cast -fdebug-prefix-map=clang/test/CodeGen/= -fno-ident -fdebug-compilation-dir=clang/test/CodeGen -debug-info-kind=limited -emit-llvm -o - clang/test/CodeGen/cfi-check-fail-debuginfo.c
1.    <eof> parser at end of file
2.    Per-file LLVM IR generation
 #0 0x0000000105c88648 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/local/bin/clang-21+0x10550c648)
 #1 0x0000000105c88c28 PrintStackTraceSignalHandler(void*) (/usr/local/bin/clang-21+0x10550cc28)
 #2 0x0000000105c86950 llvm::sys::RunSignalHandlers() (/usr/local/bin/clang-21+0x10550a950)
 #3 0x0000000105c89570 SignalHandler(int, __siginfo*, void*) (/usr/local/bin/clang-21+0x10550d570)
 #4 0x0000000190dfe584 (/usr/lib/system/libsystem_platform.dylib+0x18047a584)
 #5 0x00000001009eb0cc llvm::MDNode::getNumOperands() const (/usr/local/bin/clang-21+0x10026f0cc)
 #6 0x00000001009eb0cc llvm::MDNode::getNumOperands() const (/usr/local/bin/clang-21+0x10026f0cc)
 #7 0x00000001009eac10 llvm::MDNode::getOperand(unsigned int) const (/usr/local/bin/clang-21+0x10026ec10)
 #8 0x0000000101ee1aec llvm::DILocation::getRawScope() const (/usr/local/bin/clang-21+0x101765aec)
 #9 0x0000000101ee1a44 llvm::DILocation::getScope() const (/usr/local/bin/clang-21+0x101765a44)
#10 0x0000000101eddecc llvm::DILocation::getFile() const (/usr/local/bin/clang-21+0x101761ecc)
#11 0x000000010673a8c4 clang::CodeGen::CGDebugInfo::CreateSyntheticInlineAt(llvm::DebugLoc, llvm::StringRef) (/usr/local/bin/clang-21+0x105fbe8c4)
#12 0x000000010673aa44 clang::CodeGen::CGDebugInfo::CreateTrapFailureMessageFor(llvm::DebugLoc, llvm::StringRef, llvm::StringRef) (/usr/local/bin/clang-21+0x105fbea44)
#13 0x0000000106974838 clang::CodeGen::CodeGenFunction::EmitTrapCheck(llvm::Value*, clang::CodeGen::SanitizerHandler, bool, llvm::StringRef, llvm::StringRef) (/usr/local/bin/clang-21+0x1061f8838)
#14 0x0000000106975f7c clang::CodeGen::CodeGenFunction::EmitCfiCheckFail() (/usr/local/bin/clang-21+0x1061f9f7c)
#15 0x0000000106d7c138 clang::CodeGen::CodeGenModule::Release() (/usr/local/bin/clang-21+0x106600138)
#16 0x0000000106fabea8 (anonymous namespace)::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) (/usr/local/bin/clang-21+0x10682fea8)
#17 0x0000000106d412e4 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/usr/local/bin/clang-21+0x1065c52e4)
#18 0x00000001094ea118 clang::ParseAST(clang::Sema&, bool, bool) (/usr/local/bin/clang-21+0x108d6e118)
#19 0x00000001078c808c clang::ASTFrontendAction::ExecuteAction() (/usr/local/bin/clang-21+0x10714c08c)
#20 0x0000000106d47344 clang::CodeGenAction::ExecuteAction() (/usr/local/bin/clang-21+0x1065cb344)
#21 0x00000001078c78fc clang::FrontendAction::Execute() (/usr/local/bin/clang-21+0x10714b8fc)
#22 0x00000001077df330 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/local/bin/clang-21+0x107063330)
#23 0x0000000107a0a724 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/local/bin/clang-21+0x10728e724)
#24 0x000000010078e178 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/local/bin/clang-21+0x100012178)
#25 0x000000010077f93c ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) (/usr/local/bin/clang-21+0x10000393c)
#26 0x000000010077e5e4 clang_main(int, char**, llvm::ToolContext const&) (/usr/local/bin/clang-21+0x1000025e4)
#27 0x00000001007bacb8 main (/usr/local/bin/clang-21+0x10003ecb8)
#28 0x0000000190a43154 
Error: Failed to update test clang/test/CodeGen/cfi-check-fail-debuginfo.c

Michael137 commented:
Yea I think what we need is to just set the current DebugLocation. We usually do this using `ApplyDebugLocation::CreateArtificial` so that the artificial function we're generating has some artificial debug-location.

When I do this in `EmitCfiCheckFail` the crash disappears and we generate reasonable looking debug-info:
```
0x00000061:   DW_TAG_subprogram
                DW_AT_name      ("__clang_trap_msg$Undefined Behavior Sanitizer$Control flow integrity check failed")
                DW_AT_artificial        (true)
                DW_AT_external  (true)
                DW_AT_inline    (DW_INL_inlined)

0x00000067:   DW_TAG_subprogram
                DW_AT_low_pc    (0x0000000000000000)
                DW_AT_high_pc   (0x0000000000000043)
                DW_AT_frame_base        (DW_OP_reg7 RSP)
                DW_AT_linkage_name      ("__cfi_check_fail")
                DW_AT_artificial        (true)
                DW_AT_external  (true)

0x0000007a:     DW_TAG_formal_parameter
                  DW_AT_location        (DW_OP_reg5 RDI)
                  DW_AT_type    (0x0000009a "void *")
                  DW_AT_artificial      (true)

0x00000081:     DW_TAG_formal_parameter
                  DW_AT_location        (DW_OP_reg4 RSI)
                  DW_AT_type    (0x0000009a "void *")
                  DW_AT_artificial      (true)

0x00000088:     DW_TAG_inlined_subroutine
                  DW_AT_abstract_origin (0x00000061 "__clang_trap_msg$Undefined Behavior Sanitizer$Control flow integrity check failed")
                  DW_AT_ranges  (0x00000030
                     [0x0000000000000029, 0x0000000000000033)
                     [0x000000000000003e, 0x0000000000000043))
                  DW_AT_call_file       ("clang/test/CodeGen/<stdin>")
                  DW_AT_call_line       (0)

0x00000093:     NULL
```

I.e., the artificial ubsan trap frame is inlined into `__cfi_check_fail`. We should probably follow-up on how this looks in LLDB. But off the top I think this should Just Work.

Michael137 commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4051,6 +4141,14 @@ void CodeGenFunction::EmitTrapCheck(llvm::Value *Checked,
+  llvm::StringRef TrapMessage = GetUBSanTrapForHandler(CheckHandlerID);
+
+  if (getDebugInfo() && !TrapMessage.empty()) {
+    TrapLocation = getDebugInfo()->CreateTrapFailureMessageFor(
```
Lets add an `assert (TrapLocation && "Artificial trap frame requires valid debug-info for its scope.")`

So if we run into the crash again this will be easier to spot

delcypher commented:
@rapidsna @patrykstefanski @hnrklssn Adding you as reviewers just so you're aware that this change is being worked on and that it will likely impact the downstream Swift fork of Clang.

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-float-cast-overflow.c:
```diff
@@ -0,0 +1,10 @@
+
+// CHECK: call void @llvm.ubsantrap(i8 5) {{.*}}!dbg [[LOC:![0-9]+]]
+// CHECK: [[LOC]] = !DILocation(line: 0, scope: [[MSG:![0-9]+]], {{.+}})
+// CHECK: distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer$Floating-point to integer conversion overflowed"
```
nit: github seems unhappy w/ the lack of newline at EOF. you may want to check your editor settings, since I see this in a number of files.

anthonyhatran replied:
Added. I actually did see that in my editor but didn't know how important it was, so thanks for pointing that out

ilovepi replied:
It's not all that important, but it keeps the GitHub UI happy, and is pretty trivial to do, so usually we try to keep it that way.

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-load-invalid-value.c:
```diff
@@ -0,0 +1,15 @@
+// RUN: %clang_cc1 -triple arm64-apple-macosx14.0.0 -O0 -debug-info-kind=standalone -dwarf-version=5 \
+// RUN: -fsanitize=bool -fsanitize-trap=bool -emit-llvm %s -o - | FileCheck %s
+
+#include <stdbool.h> 
```
Same as above w/ standard headers. please address throughout the patch.

ilovepi replied:
per my other comment, this is fine since its compiler provided.

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-missing-return.cpp:
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -triple arm64-apple-macosx14.0.0 -O0 -debug-info-kind=standalone -dwarf-version=5 \
+// RUN: -fsanitize=return -fsanitize-trap=return -emit-llvm %s -o - | FileCheck %s
+
+int missing_return(int x)
+{
+    if (x > 0)
+        return x;
+}
```
nit: can we use consistent formatting?  I think you can use clang-format on these directly if you change the max line length so the comments aren't reformatted. I've done something similar in the past, but ymmv.

anthonyhatran replied:
Done

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-nullability-return.c:
```diff
@@ -0,0 +1,18 @@
+    return &x;
+}
+
+// CHECK: call void @llvm.ubsantrap(i8 15) {{.*}}!dbg [[LOC:![0-9]+]]
```
I think most of these chould have a CHECK-LABEL: that matches the function definition. If more functions are ever added, it will make the tests easier to update.

anthonyhatran replied:
Added

ilovepi reviewed (commented):
There's a large number of tests. Many of these look like they're close to an existing test ... in these cases is it possible to just use the existing test file and add checks under a prefix?

usama54321 commented on clang/test/CodeGen/cfi-icall-normalize2-debuginfo.c:
```diff
@@ -87,38 +87,42 @@ void baz(void (*fn)(int, int, int), int arg1, int arg2, int arg3) {
-// CHECK: [[META45]] = !DISubroutineType(types: [[META46:![0-9]+]])
-// CHECK: [[META46]] = !{null, [[META47:![0-9]+]], [[META14]], [[META14]], [[META14]]}
-// CHECK: [[META47]] = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: [[META48:![0-9]+]], size: 64)
+// CHECK: [[DBG27]] = !DILocation(line: 0, scope: [[META28:![0-9]+]], inlinedAt: [[DBG21]])
```
I would personally use a different prefix than DBG just to minimize the diff/make the test clearer to read, maybe something like DBGTRAP

usama54321 replied:
Since a lot of the diff here is just incrementing the suffix number of the checks for existing debug information

anthonyhatran replied:
Changed to DBGTRAP

usama54321 commented on clang/test/CodeGen/ubsan-trap-reason-add-overflow.c:
```diff
@@ -0,0 +1,10 @@
+  return a + b;
+}
+
+// CHECK: call void @llvm.ubsantrap(i8 0) {{.*}}!dbg [[LOC:![0-9]+]]
```
Why is there no line number. And the LOC/MSG matches are not being checked anywhere either

usama54321 commented on clang/test/CodeGen/ubsan-trap-reason-dynamic-type-cache-miss.cpp:
```diff
@@ -0,0 +1,23 @@
+
+// CHECK: call void @llvm.ubsantrap(i8 4) {{.*}}!dbg [[LOC:![0-9]+]]
+// CHECK: [[LOC]] = !DILocation(line: 0, scope: [[MSG:![0-9]+]], {{.+}})
+// CHECK: distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer$Dynamic type cache miss, member call made on an object whose dynamic type differs from the expected type"
```
+1

PiJoules commented:
This might be a bit orthogonal to this patch, but I think rather than using the mangled dollar scheme used by `__builtin_verbose_trap` to encode the UBSan description, it might be cleaner to instead use dwarf's `DW_AT_description` to actually store the UBSan description. I think ideally we'd have the artificial function you create just be the same as any of the typical ubsan handler functions (like `__ubsan_handle_cfi_check_fail`) and its arguments are the registers, memory locations, or values relevant to this particular ubsan check. Then lldb could just print the function with its arguments as well as a the `DW_AT_description` and from there the dev sees what they need to inspect in lldb to see what's wrong. Since the function is artificial, I think lldb or llvm-symbolizer should also be able to tell that this isn't a real function and instead point to the original function where the actual ubsan intrinsic was located.

thurstond commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -85,6 +85,96 @@ enum VariableTypeDescriptorKind : uint16_t {
+    return "Signed integer subtraction overflowed";
+
+  case SanitizerHandler::TypeMismatch:
+    return "Type mismatch in operation";
```
Suggestion: elaborate that this can be an alignment, null and/or object-size check failure

thurstond replied:
The Linux kernel also has a library that depends on enum values: https://github.com/torvalds/linux/blob/master/lib/ubsan.h
```
/*
 * ABI defined by Clang's UBSAN enum SanitizerHandler:
 * https://github.com/llvm/llvm-project/blob/release/16.x/clang/lib/CodeGen/CodeGenFunction.h#L113
 */
enum ubsan_checks {
	ubsan_add_overflow,
	ubsan_builtin_unreachable,
        ...
```

Be careful not to end up on https://www.reddit.com/r/linusrants/

anthonyhatran commented on clang/test/CodeGen/ubsan-trap-debugloc.c:
```diff
@@ -1,3 +1,4 @@
+// CASE 1
```
Had no meaning besides some use when I was fixing it, now removed.

delcypher commented:
@PiJoules 

> This might be a bit orthogonal to this patch, but I think rather than using the mangled dollar scheme used by `__builtin_verbose_trap` to encode the UBSan description, it might be cleaner to instead use dwarf's `DW_AT_description` to actually store the UBSan description. I think ideally we'd have the artificial function you create just be the same as any of the typical ubsan handler functions (like `__ubsan_handle_cfi_check_fail`) and its arguments are the registers, memory locations, or values relevant to this particular ubsan check. Then lldb could just print the function with its arguments as well as a the `DW_AT_description` and from there the dev sees what they need to inspect in lldb to see what's wrong. Since the function is artificial, I think lldb or llvm-symbolizer should also be able to tell that this isn't a real function and instead point to the original function where the actual ubsan intrinsic was located.

I wouldn't say it's orthogonal. It's a very interesting implementation idea but there are several reasons why we don't want to pursue it right now.

1. What you are proposing is significantly more complicated than what is implemented in this PR. This work being part of a GSoC project means it is time limited and we don't have time to fully implement what you are proposing. 

2. The "managed dollar scheme" used by `__builtin_verbose_trap` is an already a well established scheme for encoding trap reasons that LLDB already understands. This approach is also used by `-fbounds-safety` (not upstreamed yet), and Swift.

3. An advantage of the current scheme in this PR is that the coupling between the debugger and compiler is very weak which means it is robust against mixing compiler and debugger versions (something that unfortunately happens in practice, despite my best efforts). The coupling is "weak" compared to what you are proposing because the DWARF consumer (typically a debugger like LLDB) just needs to parse the fake function name. The downside though is the "trap reason" is limited to being a static string and thus can only include information known at compile time.

4. Your proposed scheme has a much tigher coupling between the debugger and compiler because you're proposing different function names for each UBSan trap type and arguments to the fake functions (essentially a more complicated ABI). This is much less robust against mixing debugger and compiler versions. OTOH I think being able to refer dynamic values would be quite valuable.

I think this PR is a substantial improvement over the status quo so I think we should land this (once we've addressed other reviewer comments) but we (the compiler) should reserve the right to change the implementation.

fmayer reviewed (commented):
nit: please fix the missing EOL for all the test files

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-alignment-assumption.c:
```diff
@@ -0,0 +1,14 @@
+// RUN: %clang_cc1 -triple arm64-apple-macosx14.0.0 -O0 -debug-info-kind=standalone -dwarf-version=5 \
+// RUN: -fsanitize=alignment -fsanitize-trap=alignment -emit-llvm %s -o - | FileCheck %s
+
+#include <stdint.h>
```
I wondered how these tests were passing, and realized that I had forgotten that stdint.h, stdbool.h, and stddef.h, are provided by the compiler.  you can ignore these comments. w.r.t. headers (as long as they're compiler provided ones).

anthonyhatran commented on clang/test/CodeGen/ubsan-trap-debugloc.c:
```diff
@@ -17,8 +18,11 @@ void bar(volatile int a) __attribute__((optnone)) {
 
 // With optimisations enabled the traps are merged and need to share a debug location
-// CHECK: [[LOC]] = !DILocation(line: 0
+// CHECK: [[LOC]] = !DILocation(line: 0,
```
Removed comma.

vitalybuka commented:
> @fmayer
> 
> > I would prefer this to be behind a flag (not necessarily in this PR), even if it's default turned on. People might not want synthetic frames added to their debug information and the top frame to point to the real code; also as this somewhat overlaps with the logic of `-fsanitize-annotate-debug-info`, with a flag people can choose which one they want.
> 
> Do you have a concrete use case where it would be problematic to emit the trap reasons in debug info as the patch does? We are not opposed to adding a flag that allows it to be disabled (e.g. `-fno-sanitize-annotated-traps` or something similar) provided the trap reasons are emitted by default. That being said we don't want to implement things that won't be used. What use case(s) do you have in mind?

I am sure it will break some users how have some stack trace analyzers. Flag will unblock compiler upgrade for them.
So having a flag sounds reasonable to me. I don't have opinion what should be a default for that flag.

In this patch probaly (copt<> "defaut is off") is OK, and in followup patch copt -> clang flag.

Given that it's done for fsanitize-annotate-debug-info, should be easy, just "replay" that patch. However that patch includes per-sanitizer granularity, which may be not needed?

vitalybuka commented:
What is debug info size impact?

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -85,6 +85,96 @@ enum VariableTypeDescriptorKind : uint16_t {
 //                        Miscellaneous Helper Methods
 //===--------------------------------------------------------------------===//
 
+static llvm::StringRef GetUBSanTrapForHandler(SanitizerHandler ID) {
```
Yes but with the caveat that  `SanitizerHandler::BoundsSafety` will need special handling because we don't want to emit a trap reason for that one because that's the value reserved for `-fbounds-safety` that handles it's trap reasons differently from the rest of UBSan.

thurstond commented:
> Complements [this feature](https://github.com/llvm/llvm-project/pull/141997); adds a synthetic inline frame named __clang_trap_msg$<Category>$<TrapMessage> in debug info. [stub, since I'm not 100% sure on this and kind of just repeats what's already been stated in the first paragraph].

AFAICS the key differences are:
- -fsanitize-annotate-debug-info (tries to) annotate all the UBSan-added instructions, not just the final (trap) instruction; this makes it useful for profiling. It works with all the UBSan modes (trap, min-rt, full rt). Additionally, it will use the precise SanitizerOrdinal name if available, only falling back to the SanitizerHandler name if it is ambiguous.
- trap reasons annotates only the trap instruction and always uses the SanitizerHandler information. The benefits are it has more informative messages and it integrates well with LLDB's frame recognizer.

(I'm not suggesting you need to put all this in the commit message; this comment is largely to +1 that these features are complementary.)

Michael137 commented:
> What is debug info size impact?

Realistically this will add a few more abbreviations into `.debug_abbrev` and only a few extra bytes per-UBSAN trap (abbreviation code + 1 ULEB128 for the index into the string offset table) into `.debug_info`. The rest of the `DW_TAG_subprogram` is encoded in the abbreviation for that fake frame. But I don't have a good intuition for how many UBSAN traps get emitted for a large codebase. So @anthonyhatran try this on a bootstrapped Clang build

anthonyhatran commented:
> What is debug info size impact?

Using bloaty to compare 2 builds of clang (one with built with my feature enabled, one with it disabled) from the same source commit:
```
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +0.3% +6.01Mi  +0.3% +6.01Mi    ,__debug_info
  +2.0% +2.26Mi  [ = ]       0    [Unmapped]
  +1.2% +1.35Mi  +1.2% +1.35Mi    ,__apple_names
  +0.0% +1.01Mi  +0.0% +1.01Mi    ,__debug_str
  +0.8%  +636Ki  +0.8%  +635Ki    ,__debug_line
  +0.4%  +161Ki  +0.4%  +161Ki    ,__debug_ranges
  +0.4% +47.9Ki  +0.4% +47.9Ki    ,__debug_abbrev
  +0.0%     +14  +0.0%     +14    ,__apple_types
  [ = ]       0  +0.0%      +8    ,__common
  [ = ]       0  +7.1%      +4    ,__thread_bss
  -0.0%      -4  -0.0%      -4    ,__const
  -0.0% -1.27Ki  -0.0% -1.27Ki    ,__cstring
  +0.2% +11.5Mi  +0.1% +9.19Mi    TOTAL
```

Michael137 commented on clang/test/CodeGen/ubsan-trap-reason-crash.cpp:
```diff
@@ -0,0 +1,22 @@
+// FIXME: Currently, a nullptr check is in place before emitting a UBSan trap reason inside 
+// CodeGenFunction::EmitTrapCheck. This is disadvantageous because we may not emit a trap 
+// message in all cases where a trap message should be emitted. The check was added to prevent
+// a crash caused by a call to EmitTypeCheck in CodeGenFunction::StartFunction. The TypeMismatch 
+// handler is triggered and tries to attach itself to the debug-info location, but StartFunction 
+// only sets the debug-info location after the prologue code-gen is done resulting in a crash.
```
Don't think we need to include all these details. Lets rephrase it to:
```suggestion**
// FIXME: We should emit a trap message for this case too. But sometimes Clang will emit a ubsan trap into the prologue of a function, at which point the debug-info locations haven't been set up yet and can't hook up our artificial inline frame.
```

anthonyhatran replied:
Rephrased

Michael137 reviewed (approved):
LGTM

(just left 1 comment regarding wording of FIXME)

shafik commented on clang/lib/CodeGen/SanitizerHandler.h:
```diff
@@ -14,35 +14,66 @@
-  SANITIZER_CHECK(AlignmentAssumption, alignment_assumption, 0)                \
-  SANITIZER_CHECK(VLABoundNotPositive, vla_bound_not_positive, 0)              \
-  SANITIZER_CHECK(BoundsSafety, bounds_safety, 0)
+  SANITIZER_CHECK(AddOverflow, add_overflow, 0, "Integer addition overflowed") \
```
Is this singed, unsigned or both? UBSan has both but I don't see two versions here, it is curious there is a sub overflow, which is not specifically delineated via `-fsanitize`, may be worth clarifying.

delcypher reviewed (changes requested):
This looking really good. I've added a few minor comments (early return in GetUBSanTrapForHandler and filing an issue for a `FIXME`) that should ideally be addressed.

I'd also like you to add brief details of this feature to Clang's release notes so it's more likely people will discover your feature. The file to modify is `clang/docs/ReleaseNotes.rst`.

I'd suggest to

1. Under `Non-comprehensive list of changes in this release` add a bullet point with something like

* Trapping UBSan (e.g. `-fsanitize-trap=undefined`) now emits a string describing the reason for trapping into the generated debug info. This feature allows debuggers (e.g. LLDB) to display the reason for trapping if the trap is reached. The string is currently encoded in the debug info as an artificial frame that claims to be inlined at the trap location. The function used for the artificial frame is an artificial function whose name encodes the reason for trapping. The encoding used is currently the same as `__builtin_verbose_trap` but might change in the future.  This feature is enabled by default but can be disabled by compiling with `-fno-sanitize-annotate-debug-info-traps`.

2. Under `New Compiler Flags` mention the new compiler flag.

* The new `-fno-sanitize-annotate-debug-info-traps` flag can be used to disable emitting the reason for trapping into the debug info when compiling with trapping UBSan (e.g. `-fsanitize-trap=undefined`j).

thurstond commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2595,6 +2595,16 @@ def fsanitize_undefined_trap_on_error
       Alias<fno_sanitize_trap_EQ>, AliasArgs<["undefined"]>;
+defm sanitize_annotate_debug_info_traps
+    : BoolFOption<
+          "sanitize-annotate-debug-info-traps",
```
`-fsanitize-annotate-debug-info-traps` could be confusing: it sounds like a subset of `-fsanitize-annotate-debug-info` (i.e., the functionality of `-fsanitize-annotate-debug-info` but only for UBSan trap mode), rather than complementary.

delcypher replied:
So I suggested the name `-fsanitize-annotate-debug-info-traps`. The reason I picked that name was

* The feature annotates the debug info on traps. The proposed flag name describes this.
* The feature is very similar conceptually to `-fsanitize-annotate-debug-info` because that feature also annotates the debug info. We did consider different names but came to the conclusion that is potentially very confusing to have two different flags that both annotate the debug info but use completely different terminology for that in that flag name.

One alternative could be something like `-fsanitize-debug-trap-reasons`. So the name here doesn't really say anything about the implementation (annotating debug info) and more about what it tries to do, i.e. help with debugging the reason for trapping.

@thurstond Do you prefer `-fsanitize-debug-trap-reasons` or have another suggestion?

thurstond replied:
I would much prefer `-fsanitize-debug-trap-reasons`.

While I agree that there is a degree of confusion from "somewhat related flags have completely different flags", I think it is a lesser evil than potentially confusing users into thinking that the flags are interconnected when they are not.

For example:
```
clang -fsanitize-annotate-debug-info=integer
      -fsanitize-annotate-debug-info-traps
```
evokes the idea to me that `-fsanitize-annotate-debug-info-traps` would only apply to integer checks, when actually they operate independently. (This is somewhat similar to the problem tackled by Anthony's WIP patch in https://github.com/llvm/llvm-project/pull/147997)

The other reason (pun not intended) that I like `-fsanitize-debug-trap-reasons` is that "trap reasons" is the most concise way to describe this patch (e.g., it's in the patch title), so it's good to have it as part of the flag name.

ilovepi commented on clang/test/CodeGen/ubsan-trap-reason-flag.c:
```diff
@@ -0,0 +1,22 @@
+
+// NO-ANNOTATE-LABEL: @add_overflow
+// NO-ANNOTATE: call void @llvm.ubsantrap(i8 0) {{.*}}!dbg [[LOC:![0-9]+]]
+// NO-ANNOTATE-NOT: __clang_trap_msg
```
nit: newline at EOF.

anthonyhatran replied:
Added.

delcypher reviewed (approved):
LGTM. Let's wait for @thurstond to respond regarding the flag name.

delcypher commented:
@anthonyhatran Oh and don't forget about https://github.com/llvm/llvm-project/pull/145967#discussion_r2229499715 as well.

anthonyhatran commented on clang/test/CodeGen/ubsan-trap-reason-crash.cpp:
```diff
@@ -0,0 +1,20 @@
+// FIXME: We should emit a trap message for this case too.
+// But sometimes Clang will emit a ubsan trap into the prologue of a function,
+// at which point the debug-info locations haven't been set up yet and
+// can't hook up our artificial inline frame.
```
Added.

anthonyhatran committed:
Emit trap reasons for UBSan checks
